File: internal_testsuites/typescript/src/test_helpers/startosis_helpers.ts
import {
    StarlarkExecutionError,
    StarlarkRunResponseLine,
    StarlarkInstruction,
    StarlarkInterpretationError,
    StarlarkValidationError
} from "kurtosis-sdk/build/core/kurtosis_core_rpc_api_bindings/api_container_service_pb";
import {Readable} from "stream";

const NEWLINE_CHAR = "\n"

export function readStreamContentUntilClosed(responseLines: Readable): Promise<[
    string,
    Array<StarlarkInstruction>,
    StarlarkInterpretationError | undefined,
    Array<StarlarkValidationError>,
    StarlarkExecutionError | undefined
]> {
    let scriptOutput = ""
    let interpretationError: StarlarkInterpretationError | undefined
    let validationErrors: Array<StarlarkValidationError> = []
    let executionError: StarlarkExecutionError | undefined
    let instructions: Array<StarlarkInstruction> = []

    return new Promise(resolve => {
        responseLines.on('data', (responseLine: StarlarkRunResponseLine) => {
            if (responseLine.getInstruction() !== undefined) {
                instructions.push(responseLine.getInstruction()!)
            } else if (responseLine.getInstructionResult() !== undefined) {
                scriptOutput += responseLine.getInstructionResult()?.getSerializedInstructionResult() + NEWLINE_CHAR
            } else if (responseLine.getError() !== undefined) {
                if (responseLine.getError()?.getInterpretationError() !== undefined) {
                    interpretationError = responseLine.getError()?.getInterpretationError()
                } else if (responseLine.getError()?.getValidationError() !== undefined) {
                    validationErrors.push(responseLine.getError()!.getInterpretationError()!)
                } else if (responseLine.getError()?.getExecutionError() !== undefined) {
                    executionError = responseLine.getError()?.getExecutionError()
                }
            }
        })
        responseLines.on('error', function () {
            if (!responseLines.destroyed) {
                responseLines.destroy();
                throw new Error("Unexpected error");
            }
        });
        responseLines.on('end', function () {
            if (!responseLines.destroyed) {
                responseLines.destroy();
                resolve([scriptOutput, instructions, interpretationError, validationErrors, executionError])
            }
        });
    })
}

File: internal_testsuites/golang/test_helpers/startosis_helpers.go
package test_helpers

import (
	"github.com/kurtosis-tech/kurtosis/api/golang/core/kurtosis_core_rpc_api_bindings"
	"strings"
)

const (
	newlineChar = "\n"
)

func ReadStreamContentUntilClosed(responseLines chan *kurtosis_core_rpc_api_bindings.StarlarkExecutionResponseLine) (string, []*kurtosis_core_rpc_api_bindings.StarlarkInstruction, *kurtosis_core_rpc_api_bindings.StarlarkInterpretationError, []*kurtosis_core_rpc_api_bindings.StarlarkValidationError, *kurtosis_core_rpc_api_bindings.StarlarkExecutionError) {
	scriptOutput := strings.Builder{}
	instructions := make([]*kurtosis_core_rpc_api_bindings.StarlarkInstruction, 0)
	var interpretationError *kurtosis_core_rpc_api_bindings.StarlarkInterpretationError
	validationErrors := make([]*kurtosis_core_rpc_api_bindings.StarlarkValidationError, 0)
	var executionError *kurtosis_core_rpc_api_bindings.StarlarkExecutionError

	for responseLine := range responseLines {
		if responseLine.GetInstruction() != nil {
			instructions = append(instructions, responseLine.GetInstruction())
		} else if responseLine.GetInstructionResult() != nil {
			scriptOutput.WriteString(responseLine.GetInstructionResult().GetSerializedInstructionResult())
			scriptOutput.WriteString(newlineChar)
		} else if responseLine.GetError() != nil {
			if responseLine.GetError().GetInterpretationError() != nil {
				interpretationError = responseLine.GetError().GetInterpretationError()
			} else if responseLine.GetError().GetValidationError() != nil {
				validationErrors = append(validationErrors, responseLine.GetError().GetValidationError())
			} else if responseLine.GetError().GetExecutionError() != nil {
				executionError = responseLine.GetError().GetExecutionError()
			}
		}
	}
	return scriptOutput.String(), instructions, interpretationError, validationErrors, executionError
}
